implement set_code_transaction for EIP-7702

joseluu opened the pull request:
### What was wrong?

Closes #315 

### How was it fixed?
a new type of transaction is implemented named set_code_transaction as well as the required validators and serializers
2 tests added
tested on pectra-devnet-6 https://pectra-devnet-6.ethpandaops.io/

#### Cute Animal Picture

![satoshi cat](https://i.seadn.io/gcs/files/b747d1d90b03a0fab74c2c3c160ca3dc.jpg?w=50&auto=format)

joseluu committed:
implement set_code_transaction for EIP-7702:

- add the newsfragment
- fix style to allow tox to pass
- fix style(3) add # noqa: E501 where needed
- resolve review remarks but making a tuple instead of a dict; added authorization signature convenience function (pr #316)
- fix style (lint), attempt to put documentation in code
- fix comment and key conversion
- fix sphinx problem
- improve documentation
- improve sphynx doc string
- taming doctest
- comment all example code so that doctest does not fail
- fix NOTE directive text
- fix payload method documentation

fselmo commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+
+class SetCodeTransaction(_TypedTransactionImplementation):
+    """
+    Represents a dynamic fee transaction access per EIP-1559.
```
This should describe the new transaction per EIP-7702

fselmo commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+    Represents a dynamic fee transaction access per EIP-1559.
+    """
+
+    # This is the second transaction to implement the EIP-2718 typed transaction.
```
```suggestion
    # This is the third transaction to implement the EIP-2718 typed transaction.
```

fselmo commented on eth_account/typed_transactions/typed_transaction.py:
```diff
@@ -108,8 +114,11 @@ def from_bytes(cls, encoded_transaction: HexBytes) -> "TypedTransaction":
+            "SetCodeTransaction",
+         ]
 
         if encoded_transaction[0] == AccessListTransaction.transaction_type:
```
Unrelated but we should probably extract `encoded_transaction[0]` as `tx_type` here by now. We re-use it enough and it makes the code more readable.

```suggestion
        tx_type = encoded_transaction[0]

        if tx_type == AccessListTransaction.transaction_type:
        ... # use tx_type for all cases here
```

fselmo commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+
+        Here, the transaction payload is:
+
+            TransactionPayload = rlp([chainId,
```
This should be updated to match the actual payload for set code tx.

kclowes commented on eth_account/typed_transactions/base.py:
```diff
@@ -74,6 +74,23 @@
         "maxFeePerGas": hexstr_if_str(to_int),
         "maxFeePerBlobGas": hexstr_if_str(to_int),
         "blobVersionedHashes": apply_formatter_to_array(hexstr_if_str(to_bytes)),
+        "authorisationList": apply_formatter_to_array(
```
What do you think about making this a tuple instead of a dict? Then we don't have to do as much gymnastics with the rlp -> rpc conversion.

joseluu replied:
accessList being array it may appear inconsistent

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+        )
+
+
+class SignedAutorizationRLP(HashableRLP):
```
```suggestion
class SignedAuthorizationRLP(HashableRLP):
```

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+        Verifies that the dictionary is well formed.
+        """
+        if blobs is not None:
+            raise ValueError("Blob data is not supported for `DynamicFeeTransaction`.")
```
```suggestion
            raise ValueError("Blob data is not supported for `SetCodeTransaction`.")
```

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+        """
+        if blobs is not None:
+            raise ValueError("Blob data is not supported for `DynamicFeeTransaction`.")
+
```
I think we want to add the line: 
`dictionary = set_transaction_type_if_needed(dictionary)` somewhere in here

kclowes replied:
I think this still holds?

fselmo replied:
I personally feel like the original API for these transactions has friction points. But it seems the way that it's meant to be used is to use the `TypedTransaction.from_dict()` which [already has this step](https://github.com/ethereum/eth-account/blob/main/eth_account/typed_transactions/typed_transaction.py#L80-L106). So this feels redundant. Though we do have certain TypedTransaction implementation classes that also [add the type](https://github.com/ethereum/eth-account/blob/main/eth_account/typed_transactions/blob_transactions/blob_transaction.py#L212-L213), and [some that don't](https://github.com/ethereum/eth-account/blob/main/eth_account/typed_transactions/dynamic_fee_transaction.py#L138-L168).

I do lean toward adding it to reduce this friction as much as possible, and I'll add it in this PR, just wanted to point out where some inconsistencies lie.

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+            and encoded_transaction[0] == cls.transaction_type
+        ):
+            raise ValueError("unexpected input")
+        # Format is (0x01 || TransactionPayload)
```
```suggestion
        # Format is (0x04 || TransactionPayload)
```

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,297 @@
+        Hashes this AccessListTransaction to prepare it for signing.
+        As per the EIP-2930 specifications, the signature is a secp256k1 signature over
+        ``keccak256(0x01 || rlp([chainId, nonce, gasPrice, gasLimit,
+        to, value, data, accessList])).``
```
needs to be updated to `SetCode`

kclowes commented on newsfragments/315.feature.rst:
```diff
@@ -0,0 +1,47 @@
+set_code_transaction for EIP-7702
+
+### Overview
```
Thank you for adding such detailed documentation! This type of documentation belongs in a file other than the newsfragment file. If it stays here, it will just end up in the release notes. You'll need to follow the pattern in docs/eth_account.typed_transactions.rst to add the methods and comments from the code. You may be able to copy/paste this into that same doc file too, otherwise maybe it should be it's own page.

kclowes commented on eth_account/typed_transactions/base.py:
```diff
@@ -74,6 +74,23 @@
+                        )
+                    ),
+                    "yParity": hexstr_if_str(to_int),
+                    "signerR": hexstr_if_str(to_int),
```
I have a slight preference to simply call this `r`, and `s`, but could be convinced otherwise.

fselmo replied:
I have a stronger preference for `r` and `s`. Sticks to the EIP and the library. I'd prefer not to deviate there.

joseluu replied:
the signed authorization is going to be circulated among the participants, so as to avoid confusion with the initial users, I made it this way. it is a matter of taste, you tell me.

fselmo replied:
I would prefer to keep `r` and `s` here

kclowes commented on eth_account/_utils/transaction_utils.py:
```diff
@@ -68,10 +75,21 @@ def transaction_rpc_to_rlp_structure(dictionary: Dict[str, Any]) -> Dict[str, An
         dictionary = dissoc(dictionary, "accessList")
         rlp_structured_access_list = _access_list_rpc_to_rlp_structure(access_list)
         dictionary = assoc(dictionary, "accessList", rlp_structured_access_list)
+        access_list = dictionary.get("accessList")
```
I don't see this getting used anywhere.

kclowes commented on tests/core/test_typed_transactions.py:
```diff
@@ -353,6 +422,8 @@
     "blob-no-explicit-type",
+    "sc-standard-case",
+    "sc-and-al",
 ]
```
Nice tests!
- It would be good if we could find at least one test with an expected hash from a trusted source like the tests repo or something similar. I'll see if I can find something here too.
- I'd like to see one more test case with more than one authorizationList too.

kclowes reviewed (commented):
Thanks @joseluu! I had a very similar branch going, but happy to take yours :) It looks like you're using both `Authorization` and `Authorisation`. If it's all right with you, I would prefer to standardize on `Authorization`, because that's how it's written in the EIP.

fselmo reviewed (commented):
Had a quick initial pass here. I had similar comments to @kclowes. We should search + replace for `Authorisation` -> `Authorization` and `authorisation` -> `authorization`. Left some other comments inline but thanks for getting this going!

joseluu committed:
fix test_mnemonic as it was miscounting the accentuated characters:

- fix logic that was reversed: when testing for chinese trad, there is a chance that all words are simplified
- test BIP39 expansion starting at 4 characters rather than 5
- better fix mnenonic testing and unicode normalization of the BIP39 words

fselmo committed:
7702 minor cleanup:

- Pull ``sign_authorization`` into the local account API.
- Remove unnecessary sede, use existing ``SignedAuthorization`` as sede.
- Remove extraneous comments, tidy up code.
- Add additional tests

- Use a pydantic model for ``SignedSetCodeAuthorization`` and give this
  some JSON-RPC attention. In the future, we may be able to use these properties
  directly to serialize on web3.py as we want, but for now add a ``as_rpc_object``
  method to the pydantic model.

fselmo committed:
Generate the authority, rather than providing a static value, for better safety

pacrob commented on eth_account/account.py:
```diff
@@ -1043,3 +1054,99 @@ def sign_typed_data(
+        .. NOTE::
+            You need to get signed one or more signed authorizations from an EOA willing to have a smart contract code associated with the EOA, this the essence of EIP-7702
+
+            in the code below in the variable authorization_to_sign:
```
```suggestion
            in the code below in the variable `auth`:
```

kclowes replied:
this is `rst`, so I think you'll want \`\`auth``. Same with the below

pacrob commented on eth_account/account.py:
```diff
@@ -1043,3 +1054,99 @@ def sign_typed_data(
+            You need to get signed one or more signed authorizations from an EOA willing to have a smart contract code associated with the EOA, this the essence of EIP-7702
+
+            in the code below in the variable authorization_to_sign:
+            - chainId is the chain id of the chain where the EOA is located or 0 if the authorization is for all chains
```
```suggestion
            - `chainId` is the chain id of the chain where the EOA is located or 0 if the authorization is for all chains
```

pacrob commented on eth_account/account.py:
```diff
@@ -1043,3 +1054,99 @@ def sign_typed_data(
+
+            in the code below in the variable authorization_to_sign:
+            - chainId is the chain id of the chain where the EOA is located or 0 if the authorization is for all chains
+            - address is the address of the smart contract code to be associated with the EOA, the address format is bytes
```
```suggestion
            - `address` is the address of the smart contract code to be associated with the EOA, the address format is bytes
```

pacrob commented on eth_account/account.py:
```diff
@@ -1043,3 +1054,99 @@ def sign_typed_data(
+            in the code below in the variable authorization_to_sign:
+            - chainId is the chain id of the chain where the EOA is located or 0 if the authorization is for all chains
+            - address is the address of the smart contract code to be associated with the EOA, the address format is bytes
+            - nonce is the nonce of the EOA, it is used to prevent replay attacks
```
```suggestion
            - `nonce` is the nonce of the EOA, it is used to prevent replay attacks
```

kclowes commented on eth_account/account.py:
```diff
@@ -1043,3 +1054,99 @@ def sign_typed_data(
+        .. NOTE::
+            You need to get signed one or more signed authorizations from an EOA willing to have a smart contract code associated with the EOA, this the essence of EIP-7702
+
+            in the code below in the variable authorization_to_sign:
```
You need a new line here to get the bullet list to render correctly:

```suggestion
            in the code below in the variable authorization_to_sign:
            
```

kclowes commented on eth_account/account.py:
```diff
@@ -1043,3 +1054,99 @@ def sign_typed_data(
+            ...     "to": "0x09616C3d61b3331fc4109a9E41a8BDB7d9776609",
+            ...     "value": "0x5af3107a4000",
+            ...     "accessList": (),
+            ...     "authorizationList": [auth_signed.as_rpc_object()],
```
This feels clunky. Would it be a huge pain to handle this on the other side so the user can just pass `auth_signed`?

fselmo replied:
I currently have [an issue open](https://github.com/ethereum/web3.py/issues/3650) to be able to deprecate this and handle it implicitly in `web3.py`. I'm not really sure how else it can be "not clunky". Open to ideas. I'd rather have the information that is already there in this object and handle the serialization as JSON-RPC. 

It's not even as simple as removing the authorization address, the signature, and the hash from this object since JSON-RPC uses camelCase. We'd have to completely conform to `camelCase` within the Python class, which we moved away from for `SignedTransaction` (`rawTransaction` -> `raw_transaction`).

fselmo replied:
Come to think of it, I think this probably needs to be addressed in the `from_dict()` of the `SetCodeTransaction`. I'm going to see how to best structure this but it may very well be that to sign the transaction in the first place, it always passes through `from_dict`.

fselmo replied:
Clunky it was indeed. Turns out my last comment was correct. We never really need to get all the way over to `web3.py` for this conversion, it should be happening inside `from_dict()`. I updated it here and added tests to make sure that the Pydantic classes are properly serialized as JSON-RPC objects.

kclowes commented on tests/core/test_mnemonic.py:
```diff
@@ -136,7 +143,7 @@ def test_generation_with_string(lang, num_words):
     # valid simplified chinese characters. In that scenario, the detection
     # algorithm will assume simplified.
-    if lang == "chinese_simplified":
+    if lang == "chinese_traditional":
```
I'm curious to know the reasoning behind changing this?

joseluu replied:
If actual langage (variable lang)  is "chinese_traditional"  and if the previous code were active, the line 149 may fail
149:     assert Mnemonic.detect_language(mnemonic) == Language(lang)

This is because, as noted, chinese_traditional contains simplified characters, it may therefore happen that the detect_langage function returns "chinese_simplified" when it is not the case, leading to assert failure at line 139
This case happened to me while executing the tests

kclowes replied:
Thank you for the clarification. That makes sense to me!

fselmo replied:
I also ended up fixing the enum case. I ran 1000 tests of traditional and simplified and I was still getting some errors. I ended up changing the test to more explicitly test that if there is a detect_language failure it is only for this one specific case and it's more clear for a reader of the code now.

I tested this against 1000 of each many times over and this seems to be working well.

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,282 @@
+        "value": 0,
+        "data": b"",
+        "accessList": [],
+        "authorizationList": [],
```
It looks like sometimes we allow a List and sometimes we allow a Tuple for authorizationList, and we should make sure that we're consciously allowing both. I don't have a strong preference for either allowing both or picking one, but just wanted to flag it.

kclowes commented on eth_account/signers/local.py:
```diff
@@ -158,3 +158,9 @@ def sign_typed_data(
             ),
         )
+
+    def sign_authorization(self, authorization: Dict[str, Any]) -> SignedMessage:
```
Is there a use case where we'd want to pass an authorization list here? For example either a new method something like `sign_authorization_list` or `sign_authorizations(self, authorizations: List[Dict[str, Any]]) ...`

fselmo replied:
Authorizations can be signed by different accounts. I think that would abstract a fairly small step.

```py
authorization_list = [
    acct.sign_authorization(auth).as_rpc_object() for auth in (
        {"address": ..., "nonce": ..., "chainId": ...},
        {"address": ..., "nonce": ..., "chainId": ...},
    )
]
```

I lean towards keeping it focused on single auths tbh. Maybe we can see if there's demand first.

kclowes replied:
That sounds like a good plan to me! 👍

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,282 @@
+        """
+        Hashes this SetCodeTransaction to prepare it for signing.
+        As per the EIP-7702 specifications, the signature is a secp256k1 signature over
+        ``keccak256(0x04 || rlp([chainId, nonce, gasPrice, gasLimit, to, value, data,
```
```suggestion
        ``keccak256(0x04 || rlp([chainId, nonce, max_priority_fee_per_gas, max_fee_per_gas, gasLimit, to, value, data,
```

kclowes reviewed (approved):
I left a few comments, but nothing blocking for when you're ready to merge @fselmo!

fselmo committed:
Changes from comments on PR #316

kclowes commented on eth_account/account.py:
```diff
@@ -675,11 +686,11 @@ def sign_transaction(
         It produces signature details and the hex-encoded transaction suitable for
-        broadcast using :meth:`w3.eth.sendRawTransaction()
-        <web3.eth.Eth.sendRawTransaction>`.
+        broadcast using :meth:`w3.eth.send_raw_transactions()
```
should be:
```suggestion
        broadcast using :meth:`w3.eth.send_raw_transaction()
```

kclowes commented on eth_account/account.py:
```diff
@@ -675,11 +686,11 @@ def sign_transaction(
-        broadcast using :meth:`w3.eth.sendRawTransaction()
-        <web3.eth.Eth.sendRawTransaction>`.
+        broadcast using :meth:`w3.eth.send_raw_transactions()
+        <web3.eth.Eth.send_raw_transactions>`.
```
same here:
```suggestion   <https://web3py.readthedocs.io/en/stable/web3.eth.html#web3.eth.Eth.send_raw_transaction>.
```

kclowes commented on eth_account/account.py:
```diff
@@ -675,11 +686,11 @@ def sign_transaction(
         To create the transaction dict that calls a contract, use contract object:
-        `my_contract.functions.my_function().buildTransaction()
+        `my_contract.functions.my_function().build_transaction()
         <http://web3py.readthedocs.io/en/latest/contracts.html#methods>`_
```
This link should be: 
https://web3py.readthedocs.io/en/stable/web3.contract.html#methods

fselmo replied:
changed to a `:meth:` reference and tested it. I think that may be better so broken links don't need to be updated 👍🏼

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,289 @@
+        Hashes this SetCodeTransaction to prepare it for signing.
+        As per the EIP-7702 specifications, the signature is a secp256k1 signature over
+        ``keccak256(0x04 || rlp([chainId, nonce, max_priority_fee_per_gas,
+        max_fee_per_gas, gasLimit, gasLimit, to, value, data, accessList,
```
```suggestion
        max_fee_per_gas, gasLimit, to, value, data, accessList,
```

kclowes commented on eth_account/typed_transactions/set_code_transaction.py:
```diff
@@ -0,0 +1,289 @@
+            to, value, data,
+            accessList,
+            authorizationsList,
+            signatureYParity, signatureR, signatureS])
```
Did we decide this should be `y_parity, r, s` or am I thinking of something else?

fselmo replied:
The payload definition in the EIP makes this distinction. I think this is fine. I'd just rather not have the actual fields named `signature...` or `signer...` and have them instead be `r`, `s`, `y_parity` / `yParity`, as both the EIP and our code currently do.

kclowes commented on tests/core/test_mnemonic.py:
```diff
@@ -136,7 +143,7 @@ def test_generation_with_string(lang, num_words):
     # NOTE: Sometimes traditional chinese can return characters that are also
     # valid simplified chinese characters. In that scenario, the detection
     # algorithm will assume simplified.
```
```suggestion
    # algorithm will assume traditional.
```

kclowes reviewed (approved):
Code LGTM, mostly left doc nits!

fselmo committed:
Make mnemonic tests more robust and explicit
